executor/grant_test.go

```diff
@@ -384,3 +384,34 @@ func (s *testSuite3) TestGrantOnNonExistTable(c *C) {
 	_, err = tk.Exec("grant Select,Update on test.xx to 'genius'")
 	c.Assert(err, IsNil)
 }
+
+func (s *testSuite3) TestGrantDynamicPrivs(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("create user dyn")
+	_, err := tk.Exec("GRANT BACKUP_ADMIN ON test.* TO dyn")
+	c.Assert(terror.ErrorEqual(err, executor.ErrIllegalPrivilegeLevel), IsTrue)
+	_, err = tk.Exec("GRANT BOGUS_GRANT ON *.* TO dyn")
+	c.Assert(terror.ErrorEqual(err, executor.ErrDynamicPrivilegeNotRegistered), IsTrue)
+
+	tk.MustExec("GRANT BACKUP_Admin ON *.* TO dyn") // grant one priv
+	tk.MustQuery("SELECT * FROM mysql.global_grants WHERE `Host` = '%' AND `User` = 'dyn' ORDER BY user,host,priv,with_grant_option").Check(testkit.Rows("dyn % BACKUP_ADMIN N"))
+
+	tk.MustExec("GRANT SYSTEM_VARIABLES_ADMIN, BACKUP_ADMIN ON *.* TO dyn") // grant multiple
+	tk.MustQuery("SELECT * FROM mysql.global_grants WHERE `Host` = '%' AND `User` = 'dyn' ORDER BY user,host,priv,with_grant_option").Check(
+		testkit.Rows("dyn % BACKUP_ADMIN N", "dyn % SYSTEM_VARIABLES_ADMIN N"),
+	)
+
+	tk.MustExec("GRANT ROLE_ADMIN, BACKUP_ADMIN ON *.* TO dyn WITH GRANT OPTION") // grant multiple with GRANT option.
+	tk.MustQuery("SELECT * FROM mysql.global_grants WHERE `Host` = '%' AND `User` = 'dyn' ORDER BY user,host,priv,with_grant_option").Check(
+		testkit.Rows("dyn % BACKUP_ADMIN Y", "dyn % ROLE_ADMIN Y", "dyn % SYSTEM_VARIABLES_ADMIN N"),
+	)
+
+	tk.MustExec("GRANT SYSTEM_VARIABLES_ADMIN, Select, ROLE_ADMIN ON *.* TO dyn") // grant mixed dynamic/non dynamic
+	tk.MustQuery("SELECT Grant_Priv FROM mysql.user WHERE `Host` = '%' AND `User` = 'dyn'").Check(testkit.Rows("N"))
+	tk.MustQuery("SELECT WITH_GRANT_OPTION FROM mysql.global_grants WHERE `Host` = '%' AND `User` = 'dyn' AND Priv='SYSTEM_VARIABLES_ADMIN'").Check(testkit.Rows("N"))
+
+	tk.MustExec("GRANT CONNECTION_ADMIN, Insert ON *.* TO dyn WITH GRANT OPTION") // grant mixed dynamic/non dynamic with GRANT option.
```

Comment:
Please add the test case for `GRANT ... WITH GRANT OPTION` to 'user1', and then the 'GRANT ...' from 'user1' to 'user2'.

Reply:
The tests in `executor/grant_test.go` only test syntax (and are all performed as `root`). Privilege tests are done in `privilege/privileges/privileges_test.go`. I think this test satisfies what you are asking for though:

https://github.com/pingcap/tidb/pull/22778/files#diff-8b507d8e9d7f506b565cbedfd8e7c860aa76099537d36ed9cdeaf931d71d64c0R1189-R1193

```
	// test ROLE_ADMIN
	_, err = se.ExecuteInternal(context.Background(), "GRANT anyrolename TO otheruser")
	c.Assert(err.Error(), Equals, "[planner:1227]Access denied; you need (at least one of) the SUPER or ROLE_ADMIN privilege(s) for this operation")
	mustExec(c, rootSe, "GRANT ROLE_ADMIN ON *.* TO notsuper")
	mustExec(c, se, "GRANT anyrolename TO otheruser")
```

Reply:
I meant that it's better to have a case like:

```
(root)
CREATE USER varUser1;
CREATE USER varUser2;
CREATE USER varUser3;

GRANT SYSTEM_VARIABLES_ADMIN ON *.* TO varUser1;
GRANT SYSTEM_VARIABLES_ADMIN ON *.* TO varUser2 WITH GRANT OPTION;
```
Then,
```
(varUser1)
GRANT SYSTEM_VARIABLES_ADMIN ON *.* TO varUser3;
```
It should fail.

```
(varUser2)
GRANT SYSTEM_VARIABLES_ADMIN ON *.* TO varUser3;
```
It should succeed because of `WITH GRANT OPTION`.
```

Reply:
This has been addressed in https://github.com/pingcap/tidb/pull/22778/commits/67c042873461b8973a5d2f69aa2cc5de16a8516d

(Yes, there was a bug in it.)